File: 148. Sort List.java
/*
Problem Link: https://leetcode.com/problems/sort-list/
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

import java.util.*;

class Solution {
    public ListNode sortList(ListNode head) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);
        ListNode curr = head;
        while (curr != null) {
            pq.add(curr);
            curr = curr.next;
        }

        ListNode dummy = new ListNode();
        curr = dummy;
        while (!pq.isEmpty()) {
            ListNode node = pq.poll();
            curr.next = node;
            curr = curr.next;
        }
        curr.next = null;
        return dummy.next;
    }
}

File: Problems/148. Sort List.py
"""
Problem Link: https://leetcode.com/problems/sort-list/
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

import heapq


class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        heap_list = []
        while head:
            heap_list.append(head.val)
            head = head.next

        heapq.heapify(heap_list)
        dummy = ListNode()
        curr = dummy
        while heap_list:
            curr.next = ListNode(heapq.heappop(heap_list))
            curr = curr.next
        return dummy.next
